code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/PushTemplateHelpers.kt

```diff
@@ -0,0 +1,313 @@
+        return if ((applicationCacheDir == null)) null else (
+            (
+                applicationCacheDir
+                    .toString() + File.separator +
+                    PushTemplateConstants.CACHE_BASE_DIR
+                ) + File.separator +
+                PushTemplateConstants.PUSH_IMAGE_CACHE
+            )
+    }
+
+private val executor: ExecutorService
+    get() = ExecutorHolder.INSTANCE
+
+private object ExecutorHolder {
+    val INSTANCE: ExecutorService = Executors.newSingleThreadExecutor()
+}
+
+private class DownloadImageCallable(val url: String?) :
+    Callable<Bitmap?> {
+    override fun call(): Bitmap? {
+        val bitmap: Bitmap?
+        var connection: HttpURLConnection? = null
+        var inputStream: InputStream? = null
+        try {
+            val imageUrl = URL(url)
+            connection = imageUrl.openConnection() as HttpURLConnection
+            inputStream = connection.inputStream
+            bitmap = BitmapFactory.decodeStream(inputStream)
+        } catch (e: IOException) {
+            Log.warning(
+                PushTemplateConstants.LOG_TAG,
+                SELF_TAG,
+                "Failed to download push notification image from url ($url). Exception: ${e.localizedMessage}"
+            )
+            return null
+        } finally {
+            if (inputStream != null) {
+                try {
+                    inputStream.close()
+                } catch (e: IOException) {
+                    Log.warning(
+                        PushTemplateConstants.LOG_TAG,
+                        SELF_TAG,
+                        "IOException during closing Input stream while push notification image from url ($url). Exception: ${e.localizedMessage}"
+
+                    )
+                }
+            }
+            connection?.disconnect()
+        }
+        Log.trace(
+            PushTemplateConstants.LOG_TAG,
+            SELF_TAG,
+            "Downloaded push notification image from url ($url)"
+        )
+        return bitmap
+    }
+}
+
+/**
+ * Downloads an image using the provided uri `String`. Prior to downloading, the image uri
+ * is used to retrieve a [CacheResult] containing a previously cached image. If no cache
+ * result is returned, a call to [download] is made to download then cache the image.
+ *
+ * If a valid cache result is returned then no image is downloaded. Instead, a `Bitmap`
+ * is created from the cache result and returned by this method.
+ *
+ * @param cacheService the AEPSDK [CacheService] to use for caching or retrieving
+ * downloaded image assets
+ * @param uri [String] containing an image asset url
+ * @return [Bitmap] containing the image referenced by the `String` uri
+ */
+internal fun downloadImage(cacheService: CacheService, uri: String?): Bitmap? {
+    if (assetCacheLocation.isNullOrEmpty() || uri.isNullOrEmpty()) {
+        return null
+    }
+    val cacheResult = cacheService[assetCacheLocation!!, uri]
```

Comment:
Strong assert `assetCacheLocation!!` is not needed since it is already being done in the previous line.

Reply:
the strong assert is used as assetCacheLocation has an open getter

Reply:
I recommend fetching the cache location when required.

Reply:
sounds good, will update

Reply:
updated this to be a utility function named `getAssetCacheLocation` instead